src/main/java/org/zalando/baigan/proxy/ConfigurationBeanDefinitionRegistrar.java

```diff
@@ -75,25 +78,61 @@ public void registerBeanDefinitions(
+                    final GenericBeanDefinition genericDefinition = (GenericBeanDefinition) definition;
+                    registerAsBean(registry, genericDefinition);
+                } else {
+                    throw new IllegalStateException(
+                        String.format(
+                            "Unable to read required metadata of configuration candidate [%s]",
+                            definition
+                        )
+                    );
+                }
+            }
+        }
+    }
 
-            final Set<Class<?>> configServiceInterfaces = new Reflections(
-                    singlePackage).getTypesAnnotatedWith(BaiganConfig.class);
+    private void registerAsBean(final BeanDefinitionRegistry registry, final GenericBeanDefinition genericDefinition) {
+        try {
+            final Class<?> interfaceToImplement = genericDefinition.resolveBeanClass(
+                registry.getClass().getClassLoader()
+            );
+            registerAsBean(registry, interfaceToImplement);
+        } catch (final ClassNotFoundException e) {
+            throw new IllegalStateException("Unable to register annotated interface as configuration bean", e);
+        }
+    }
 
-            for (final Class<?> interfaceToImplement : configServiceInterfaces) {
-                final BeanDefinitionBuilder builder = BeanDefinitionBuilder
-                        .genericBeanDefinition(
-                                ConfigurationServiceBeanFactory.class);
-                builder.addPropertyValue("candidateInterface",
-                        interfaceToImplement);
+    private void registerAsBean(final BeanDefinitionRegistry registry, final Class<?> interfaceToImplement) {
+        final BeanDefinitionBuilder builder = BeanDefinitionBuilder.genericBeanDefinition(
+            ConfigurationServiceBeanFactory.class
+        );
+        builder.addPropertyValue("candidateInterface", interfaceToImplement);
 
-                final String factoryBeanName = interfaceToImplement.getName()
-                        + "BaiganProxyConfigurationFactoryBean";
-                registry.registerBeanDefinition(factoryBeanName,
-                        builder.getBeanDefinition());
-            }
+        final String factoryBeanName = interfaceToImplement.getName() + "BaiganProxyConfigurationFactoryBean";
+        registry.registerBeanDefinition(factoryBeanName, builder.getBeanDefinition());
+    }
+
+    private static class ClassPathScanningCandidateInterfaceProvider
+        extends ClassPathScanningCandidateComponentProvider {
+
+        public ClassPathScanningCandidateInterfaceProvider() {
+            super(false);
+        }
 
+        @Override
+        protected boolean isCandidateComponent(AnnotatedBeanDefinition beanDefinition) {
+            final AnnotationMetadata metadata = beanDefinition.getMetadata();
+            return metadata.isIndependent() && metadata.isInterface();
```

Comment:
What is our approach for test coverage? There is a test `ConfigurationBeanDefinitionRegistrarTest` and it's still green. But it's not testing all branches.

Reply:
I don't think there is any aligned approach for testing yet, looking thru the existing code base it seems pretty random in terms of coverage (parts are not covered at all, others are multi-covered). It might make sense to expand on https://github.com/zalando-stups/baigan-config/issues/73 and tackle testing more holistically